packages/protocol/contracts/L1/libs/LibAuction.sol

```diff
@@ -0,0 +1,149 @@
+// SPDX-License-Identifier: MIT
+//  _____     _ _         _         _
+// |_   _|_ _(_) |_____  | |   __ _| |__ ___
+//   | |/ _` | | / / _ \ | |__/ _` | '_ (_-<
+//   |_|\__,_|_|_\_\___/ |____\__,_|_.__/__/
+
+pragma solidity ^0.8.18;
+
+import {AddressResolver} from "../../common/AddressResolver.sol";
+import {LibAddress} from "../../libs/LibAddress.sol";
+import {LibUtils} from "./LibUtils.sol";
+import {TaikoData} from "../../L1/TaikoData.sol";
+import {TaikoToken} from "../TaikoToken.sol";
+
+library LibAuction {
+    using LibAddress for address;
+
+    event Bid(
+        uint256 indexed id,
+        address claimer,
+        uint256 claimedAt,
+        uint256 deposit,
+        uint256 minFeePerGasInWei
+    );
+
+    event BidDepositRefunded(
+        uint256 indexed batchStartId, address claimer, uint256 refundedAt, uint256 refund
+    );
+
+    error L1_BLOCK_ID();
+    error L1_BID_AUCTION_CLOSED();
+    error L1_BID_NOT_ACCEPTABLE();
+    error L1_ID_NOT_START_OF_A_BATCH();
+
+    function bidForBatch(
+        TaikoData.State storage state,
+        AddressResolver resolver,
+        TaikoData.Config memory config,
+        uint256 batchStartBlockId,
+        uint256 minFeePerGasInWei
+    ) internal {
+        // If it verified already -> Revert, otherwise it would be possible
+        // to bid for future (unproposed) blocks - tho risky given the unknown costs yet.
+        if (batchStartBlockId <= state.lastVerifiedBlockId) {
+            revert L1_BLOCK_ID();
+        }
+
+        // We also need to be sure, that the blockId here is indeed a start blockId
+        if (batchStartBlockId != getBatchStartBlockId(config, batchStartBlockId)) {
+            revert L1_ID_NOT_START_OF_A_BATCH();
+        }
+
+        TaikoData.Bid memory currentBid = state.bids[batchStartBlockId];
+
+        // if a current bid exists, and the delay after the current bid exists
+        // has passed, bidding is over.
+        if (!isBiddingOpenForBlock(config, currentBid)) {
+            revert L1_BID_AUCTION_CLOSED();
+        }
+
+        TaikoToken tkoToken = TaikoToken(resolver.resolve("tko_token", false));
+
+        // If there is an existing bid already
+        // we compare this bid to the existing one first, to see if its a
+        // lower fee accepted.
+        if (currentBid.bidder != address(0)) {
```

Comment:
Always use a ring buffer please, this check is then incorrect -- we need to check if batchId matches, bidder (call it prover please) will not be reset to address(0).

Reply:
They are not sharing the same ring buffer, so why it is not possible?